### ISSUE TITLE
Reduction Operators

### ISSUE THREAD

**Issue by ShiQiu0419** [zeliu98/CloserLook3D#2]
Hi,

Thanks for your great work.

Instead of the local aggregation operators, I am curious about the following reduction operators. In your codes, the MAX or AVG seems different from the regular max/avg-pooling along neighbors, due to certain masks involved. (and MAX operators changes in pytorch and tensorflow implementation?).

Could you please explain a bit more on them especially for the masks? 

Thanks a lot.

**Comment by zeliu98** [comment]
Hi @ShiQiu0419 

This an implementation difference between tensorflow and pytorch for variable input and neighborhood. For tensorflow, we calculate the subsampling indices and neighborhood indices in dataloader following KPCONV(https://github.com/HuguesTHOMAS/KPConv). For pytorch, we do subsampling and neighborhood quering as basic operators in network. So to deal with variable input and variable neighborhood, we add paddings for points and neighborhoods. These masks indicate these paddings. And in Reduction functions, these masks are used to eliminate these paddings for more accurate results.
Hope this help!

Best,
Ze

**Comment by ShiQiu0419** [comment]
Hi @zeliu98 

Thanks for your reply. Sorry that I am not farmiliar with the implementation of KPConv.

So actually, your reduction functions aim to max/avg-pool along neighbors (like regular PointNet++ did), while extra operations in your codes only deal with some specific masks in your/KPConv settings?

**Comment by zeliu98** [comment]
Hi @ShiQiu0419 

I'm a little confused about what you mean. 
Let me take an example. For finding neighborhoods, we use ball query (set a ball radius and find all neighbor points in this ball), but the number of neighbor points is different for each center point due to different densities. So we usually set a max neighborhood number `K`. For center point which has  more than `K` neighbors, we choose the `K` nearest neighbor points; while for center point which has less than `K` neighbors, we use some padding points to fill to `K` neighbor points. But in reduction functions, we don't want to use these padding points, so we set a mask to indicate them and only use the true neighbor points. That's why we set these neighborhood masks and how they are used.

Best.
Ze

**Comment by ShiQiu0419** [comment]
Hi  @zeliu98 

Many thanks for the clarification. 

I was unclear about the "mask" or "padding points" before (I suppose there was no such mask in PointNet++..) Anyway, it makes sense to me now.  Really appreciate your great efforts on this work.
